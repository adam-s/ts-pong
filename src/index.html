<!DOCTYPE html> 
<html>
<head>
  <title>Pong!</title> 
  <meta http-equiv="Content-Type" content="text/html; charset=utf-8"/> 
  <style href="pong.css" media="screen, print" rel="stylesheet" type="text/css"></style> 
</head> 
 
<body> 

  <div id="sidebar">

    <h2>This is Pong!</h2>

    <div class='description'>
      <p>
        This is a javascript version of Pong.
      </p>
      <p>
        Press <b>1</b> for a single player game.<br>
        Press <b>2</b> for a double player game.<br>
        Press <b>0</b> to watch the computer play itself.
      </p>
      <p>
        Player 1 moves using the <b>Q</b> and <b>A</b> keys.<br>
        Player 2 moves using the <b>P</b> and <b>L</b> keys.
      </p>
      <p>
        <b>Esc</b> can be used to abandon a game.
      </p>
    </div>

    <div class='settings'>
      <label for='sound'>sound: </label>
      <input type='checkbox' id='sound'>
    </div>

    <div class='settings'>
      <label for='stats'>stats: </label>
      <input type='checkbox' id='stats' checked>
    </div>

    <div class='settings'>
      <label for='footprints'>footprints: </label>
      <input type='checkbox' id='footprints'>
    </div>

    <div class='settings'>
      <label for='predictions'>predictions: </label>
      <input type='checkbox' id='predictions'>
    </div>

  </div>

  <canvas id="game">
    <div id="unsupported">
      Sorry, this example cannot be run because your browser does not support the &lt;canvas&gt; element
    </div> 
  </canvas>
</body> 
</html>
