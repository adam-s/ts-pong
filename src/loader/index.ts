export * from './Loader';
export * from './Resource';
export * from './ImageSource';
export * from './AudioSource';
export * from './AudioContext';
export * from './WebAudioInstance';
