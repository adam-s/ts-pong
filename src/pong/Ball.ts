/* eslint-disable @typescript-eslint/no-empty-function */
import { Engine } from '../engine';
import { accelerate, Drawable, intercept, Updatable } from '../utils';
import { Paddle } from '../pong';
import { Colors } from './Pong';

export interface BallOptions {
  ballRadius: number;
  width: number;
  height: number;
  wallWidth: number;
  ballSpeed: number;
  ballAccel: number;
  footprints: boolean;
}

export class Ball implements Updatable, Drawable {
  private _x = 0;

  private _y = 0;

  private _dx: number;

  private _dy: number;

  private _radius: number;

  private minX: number;

  private maxX: number;

  private minY: number;

  private maxY: number;

  private speed: number;

  private accel: number;

  // private top: number;

  // private bottom: number;

  private _left: number;

  private _right: number;

  private dxChanged = false;

  private dyChanged = false;

  private footprints: Array<{ x: number; y: number }> = [];

  private footprintCount = 0;

  public showFootprints = false;

  constructor(options: BallOptions) {
    const {
      ballRadius,
      width,
      height,
      wallWidth,
      ballSpeed,
      ballAccel,
      footprints,
    } = options;
    this.minX = this._radius = ballRadius;
    this.maxX = width - this._radius;
    this.minY = wallWidth + this._radius;
    this.maxY = height - wallWidth - this._radius;
    this.speed = (this.maxX - this.minX) / ballSpeed;
    this.accel = ballAccel;
    this.showFootprints = footprints;
  }

  get x() {
    return this._x;
  }

  get y() {
    return this._y;
  }

  get dx() {
    return this._dx;
  }

  get dy() {
    return this._dy;
  }

  get radius() {
    return this._radius;
  }

  get left() {
    return this._left;
  }

  get right() {
    return this._right;
  }

  reset(playerNumber: number | null) {
    this.footprints = [];
    this.setpos(
      playerNumber === 1 ? this.maxX : this.minX,
      Engine.random(this.minY, this.maxY)
    );
    this.setdir(playerNumber === 1 ? -this.speed : this.speed, this.speed);
  }

  setpos(x: number, y: number) {
    this._x = x;
    this._y = y;
    this._left = this._x - this.radius;
    // this.top = this._y - this.radius;
    this._right = this._x + this.radius;
    // this.bottom = this._y + this.radius;
  }

  setdir(dx: number, dy: number) {
    this.dxChanged = this._dx < 0 !== dx < 0; // did horizontal direction change
    this.dyChanged = this._dy < 0 !== dy < 0; // did vertical direction change
    this._dx = dx;
    this._dy = dy;
  }

  footprint() {
    if (this.showFootprints) {
      if (!this.footprintCount || this.dxChanged || this.dyChanged) {
        this.footprints.push({ x: this.x, y: this.y });
        if (this.footprints.length > 50) this.footprints.shift();
        this.footprintCount = 5;
      } else {
        this.footprintCount--;
      }
    }
  }

  intercept(
    rect: { right: number; left: number; top: number; bottom: number },
    nx: number,
    ny: number
  ) {
    let pt;
    // Moving left
    if (nx < 0) {
      pt = intercept({
        x1: this.x, // x coord on the right side of ball segment
        y1: this.y, // y coord on the right side of ball segment
        x2: this.x + nx, // x coord where ball will be painted moving left
        y2: this.y + ny, // y coord where the ball will be painted moving left
        x3: rect.right + this.radius,
        y3: rect.top - this.radius,
        x4: rect.right + this.radius,
        y4: rect.bottom + this.radius,
        d: 'right',
      });
    } else if (nx > 0) {
      pt = intercept({
        x1: this.x,
        y1: this.y,
        x2: this.x + nx,
        y2: this.y + ny,
        x3: rect.left - this.radius,
        y3: rect.top - this.radius,
        x4: rect.left - this.radius,
        y4: rect.bottom + this.radius,
        d: 'left',
      });
    }
    if (!pt) {
      if (ny < 0) {
        pt = intercept({
          x1: this.x,
          y1: this.y,
          x2: this.x + nx,
          y2: this.y + ny,
          x3: rect.left - this.radius,
          y3: rect.bottom + this.radius,
          x4: rect.right + this.radius,
          y4: rect.bottom + this.radius,
          d: 'bottom',
        });
      } else if (ny > 0) {
        pt = intercept({
          x1: this.x,
          y1: this.y,
          x2: this.x + nx,
          y2: this.y + ny,
          x3: rect.left - this.radius,
          y3: rect.top - this.radius,
          x4: rect.right + this.radius,
          y4: rect.top - this.radius,
          d: 'top',
        });
      }
    }
    return pt;
  }

  update(dt: number, leftPaddle: Paddle, rightPaddle: Paddle) {
    const pos = accelerate({
      x: this.x,
      y: this.y,
      dx: this.dx,
      dy: this.dy,
      accel: this.accel,
      dt,
    });

    // Reverse direction (up) if ball hits bottom wall
    if (pos.dy > 0 && pos.y > this.maxY) {
      pos.y = this.maxY; // Set the y position to the top of the bottom wal
      pos.dy = -pos.dy; // Reverse direction
      // Reverse direction (down) if ball hits top wall
    } else if (pos.dy < 0 && pos.y < this.minY) {
      pos.y = this.minY; // Set the y position to the bottom of the top wall
      pos.dy = -pos.dy; // Reverse direction
    }

    // Choose paddle based on direction of ball. Negative from right to left is left paddle
    const paddle = pos.dx < 0 ? leftPaddle : rightPaddle;
    // Test if the ball will intercept the current paddle
    const pt = this.intercept(paddle, pos.nx, pos.ny);
    // If there is an interception with a paddle set the position to the point of intercept
    // Switch directions based on left / right or top / bottom
    if (pt) {
      switch (pt.d) {
        case 'left':
        case 'right':
          pos.x = pt.x;
          pos.dx = -pos.dx;
          break;
        case 'top':
        case 'bottom':
          pos.y = pt.y;
          pos.dy = -pos.dy;
          break;
      }

      // add/remove spin based on paddle direction
      if (paddle.up) pos.dy = pos.dy * (pos.dy < 0 ? 0.5 : 1.5);
      else if (paddle.down) pos.dy = pos.dy * (pos.dy > 0 ? 0.5 : 1.5);
    }

    this.setpos(pos.x, pos.y);
    this.setdir(pos.dx, pos.dy);
    this.footprint();
  }

  draw(ctx: CanvasRenderingContext2D) {
    const side = this.radius * 2;
    ctx.fillStyle = Colors.ball;
    ctx.fillRect(this.x - this.radius, this.y - this.radius, side, side);
    if (this.showFootprints) {
      const max = this.footprints.length;
      ctx.strokeStyle = Colors.footprint;
      for (let n = 0; n < max; n++)
        ctx.strokeRect(
          this.footprints[n].x - this.radius,
          this.footprints[n].y - this.radius,
          side,
          side
        );
    }
  }
}
