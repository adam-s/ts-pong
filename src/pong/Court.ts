import { Drawable, BlockElement } from '../utils';
import { Colors } from './Pong';

interface CourtOptions {
  width: number;
  height: number;
  wallWidth: number;
}

const DIGITS = [
  [1, 1, 1, 0, 1, 1, 1], // 0
  [0, 0, 1, 0, 0, 1, 0], // 1
  [1, 0, 1, 1, 1, 0, 1], // 2
  [1, 0, 1, 1, 0, 1, 1], // 3
  [0, 1, 1, 1, 0, 1, 0], // 4
  [1, 1, 0, 1, 0, 1, 1], // 5
  [1, 1, 0, 1, 1, 1, 1], // 6
  [1, 0, 1, 0, 0, 1, 0], // 7
  [1, 1, 1, 1, 1, 1, 1], // 8
  [1, 1, 1, 1, 0, 1, 0], // 9
] as const;

export class Court implements Drawable {
  private wallWidth: number;

  private walls: BlockElement[] = [];

  private score1: BlockElement;

  private score2: BlockElement;

  constructor(options: CourtOptions) {
    const { width, height, wallWidth } = options;
    this.wallWidth = wallWidth;

    // Top wall
    this.walls.push({ x: 0, y: 0, width, height: wallWidth });

    // Bottom wall
    this.walls.push({ x: 0, y: height - wallWidth, width, height: wallWidth });

    // Draw vertical dashed line halfway
    const innerHeight = height / (wallWidth * 2);
    for (let n = 0; n < innerHeight; n++) {
      this.walls.push({
        x: width / 2 - wallWidth / 2,
        y: wallWidth / 2 + wallWidth * 2 * n,
        width: wallWidth,
        height: wallWidth,
      });
    }

    const scoreWidth = 3 * wallWidth;
    const scoreHeight = 4 * wallWidth;

    this.score1 = {
      x: 0.5 + width / 2 - 1.5 * wallWidth - scoreWidth,
      y: 2 * wallWidth,
      width: scoreWidth,
      height: scoreHeight,
    };

    this.score2 = {
      x: 0.5 + width / 2 + 1.5 * wallWidth,
      y: 2 * wallWidth,
      width: scoreWidth,
      height: scoreHeight,
    };
  }

  draw(
    ctx: CanvasRenderingContext2D,
    scorePlayer1: number,
    scorePlayer2: number
  ) {
    ctx.fillStyle = Colors.walls;
    for (let i = 0; i < this.walls.length; i++) {
      const { x, y, width, height } = this.walls[i];
      ctx.fillRect(x, y, width, height);
    }

    this.drawDigit(ctx, scorePlayer1, this.score1);
    this.drawDigit(ctx, scorePlayer2, this.score2);
  }

  drawDigit(
    ctx: CanvasRenderingContext2D,
    score: number,
    scoreElement: BlockElement
  ) {
    const { x, y, width, height } = scoreElement;
    ctx.fillStyle = Colors.walls;

    const drawWidth = (this.wallWidth * 4) / 5;
    const drawHeight = drawWidth; // For self documentation

    const blocks = DIGITS[score];

    if (blocks[0]) {
      ctx.fillRect(x, y, width, drawWidth);
    }
    if (blocks[1]) {
      ctx.fillRect(x, y, drawWidth, height / 2);
    }
    if (blocks[2]) {
      ctx.fillRect(x + width - drawWidth, y, drawWidth, height / 2);
    }
    if (blocks[3]) {
      ctx.fillRect(x, y + height / 2 - drawHeight / 2, width, drawHeight);
    }
    if (blocks[4]) {
      ctx.fillRect(x, y + height / 2, drawWidth, height / 2);
    }
    if (blocks[5]) {
      ctx.fillRect(
        x + width - drawWidth,
        y + height / 2,
        drawWidth,
        height / 2
      );
    }
    if (blocks[6]) {
      ctx.fillRect(x, y + height - drawHeight, width, drawHeight);
    }
  }
}
