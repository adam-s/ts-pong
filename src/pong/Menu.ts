import { Drawable, ImageElement } from '../utils';
import { PlayerNumber } from './Pong';

interface MenuOptions {
  images: { [key: string]: HTMLImageElement };
  width: number;
  wallWidth: number;
}

export class Menu implements Drawable {
  private press1: ImageElement;

  private press2: ImageElement;

  private winner1: ImageElement;

  private winner2: ImageElement;

  private winner: PlayerNumber;

  constructor(options: MenuOptions) {
    const { images, width, wallWidth } = options;
    const press1 = images['press1'];
    const press2 = images['press2'];
    const winner = images['winner'];
    this.press1 = { image: press1, x: 10, y: wallWidth };
    this.press2 = {
      image: press2,
      x: width - press2.width - 10,
      y: wallWidth,
    };
    this.winner1 = {
      image: winner,
      x: width / 2 - winner.width - wallWidth,
      y: 6 * wallWidth,
    };
    this.winner2 = {
      image: winner,
      x: width / 2 + wallWidth,
      y: 6 * wallWidth,
    };
  }

  declareWinner(playerNumber: PlayerNumber) {
    this.winner = playerNumber;
  }

  draw(ctx: CanvasRenderingContext2D) {
    // Maybe wait for images to load
    ctx.drawImage(this.press1.image, this.press1.x, this.press1.y);
    ctx.drawImage(this.press2.image, this.press2.x, this.press2.y);
    if (this.winner == 0)
      ctx.drawImage(this.winner1.image, this.winner1.x, this.winner1.y);
    else if (this.winner == 1)
      ctx.drawImage(this.winner2.image, this.winner2.x, this.winner2.y);
  }
}
