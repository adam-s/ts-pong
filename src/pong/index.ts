export * from './Pong';
export * from './Ball';
export * from './Court';
export * from './Menu';
export * from './Paddle';
export * from './resources';
