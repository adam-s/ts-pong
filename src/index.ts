import './styles.scss';
import { Engine } from './engine/Engine';
import { Pong, loader } from './pong';

const pong = new Pong();
const engine = new Engine(pong, { canvasElementId: 'game' });

engine.start(loader).then(() => {
  const stats = document.getElementById('stats') as HTMLInputElement;
  const sound = document.getElementById('sound') as HTMLInputElement;
  const footprints = document.getElementById('footprints') as HTMLInputElement;
  const predictions = document.getElementById(
    'predictions'
  ) as HTMLInputElement;

  // engine.showStats(stats.checked);
  // pong.playSounds(sound.checked);
  // pong.showFootprints(footprints.checked);
  // pong.showPredictions(predictions.checked);

  stats.addEventListener('change', () => engine.showStats(stats.checked));
  sound.addEventListener('change', () => pong.playSounds(sound.checked));
  footprints.addEventListener('change', () =>
    pong.showFootprints(footprints.checked)
  );
  predictions.addEventListener('change', () =>
    pong.showPredictions(predictions.checked)
  );
});
