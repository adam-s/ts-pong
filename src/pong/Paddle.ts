import { Colors, Level, Levels } from './Pong';
import { Direction, Drawable, Updatable } from '../utils';
import { Ball } from './Ball';
import { Engine } from '../engine/Engine';

/* eslint-disable @typescript-eslint/no-empty-function */
interface PaddleOptions {
  rightHandSide?: boolean;
  paddleWidth: number;
  paddleHeight: number;
  paddleSpeed: number;
  wallWidth: number;
  height: number;
  width: number;
  showPredictions: boolean;
}

interface Predication {
  x: number;
  y: number;
  d: Direction;
  dx: number;
  dy: number;
  exactX: number;
  exactY: number;
  radius: number;
  since: number;
}

export class Paddle implements Drawable, Updatable {
  private width: number;

  private height: number;

  private paddleWidth: number;

  private paddleHeight: number;

  private minY: number;

  private maxY: number;

  private speed: number;

  private x: number;

  private y: number;

  private _left: number;

  private _right: number;

  private _top: number;

  private _bottom: number;

  private _up: number;

  private _down: number;

  private _auto = false;

  private level: Level;

  private prediction: Predication;

  public showPredications: boolean;

  constructor(options: PaddleOptions) {
    const {
      rightHandSide,
      paddleWidth,
      paddleHeight,
      paddleSpeed,
      wallWidth,
      height,
      width,
      showPredictions,
    } = options;

    this.width = width;
    this.height = height;
    this.paddleWidth = paddleWidth;
    this.paddleHeight = paddleHeight;
    this.minY = wallWidth;
    this.maxY = height - wallWidth - this.paddleHeight;
    this.speed = (this.maxY - this.minY) / paddleSpeed;
    this.setpos(
      rightHandSide ? width - paddleWidth : 0,
      this.minY + (this.maxY - this.minY) / 2
    );
    this.setdir(0);
    this.level = Levels[8];
    this.showPredications = showPredictions;
  }

  setpos(x: number, y: number) {
    this.x = x;
    this.y = y;
    this._left = this.x;
    this._right = this._left + this.paddleWidth;
    this._top = this.y;
    this._bottom = this.y + this.paddleHeight;
  }

  setdir(dy: number) {
    this._up = dy < 0 ? -dy : 0;
    this._down = dy > 0 ? dy : 0;
  }

  setAuto(on: boolean): void;
  setAuto(on: boolean, level: number): void;
  setAuto(on: boolean, level?: number) {
    if (on && !this._auto && level) {
      this._auto = true;
      this.setLevel(level);
    } else if (!on && this._auto) {
      this._auto = false;
      this.setdir(0);
    }
  }

  get right() {
    return this._right;
  }

  get left() {
    return this._left;
  }

  get top() {
    return this._top;
  }

  get bottom() {
    return this._bottom;
  }

  get up() {
    return this._up;
  }

  get down() {
    return this._down;
  }

  get auto() {
    return this._auto;
  }

  setLevel(level: number) {
    if (this._auto) this.level = Levels[level];
  }

  update(dt: number, ball: Ball) {
    if (this._auto) {
      this.ai(dt, ball);
    }

    const amount = this._down - this._up;
    if (amount !== 0) {
      let y = this.y + amount * dt * this.speed;
      if (y < this.minY) {
        y = this.minY;
      } else if (y > this.maxY) {
        y = this.maxY;
      }
      this.setpos(this.x, y);
    }
  }

  ai(dt: number, ball: Ball) {
    if (
      (ball.x < this._left && ball.dx < 0) ||
      (ball.x > this._right && ball.dx > 0)
    ) {
      this.stopMovingUp();
      this.stopMovingDown();
    } else {
      this.predict(dt, ball);
      if (this.prediction) {
        if (this.prediction.y < this._top + this.paddleHeight / 2 - 5) {
          this.stopMovingDown();
          this.moveUp();
        } else if (
          this.prediction.y >
          this._bottom - this.paddleHeight / 2 + 5
        ) {
          this.stopMovingUp();
          this.moveDown();
        } else {
          this.stopMovingUp();
          this.stopMovingDown();
        }
      }
    }
  }

  predict(dt: number, ball: Ball) {
    // only re-predict if the ball changed direction, or its been some amount of time since last prediction
    if (
      this.prediction &&
      this.prediction.dx * ball.dx > 0 &&
      this.prediction.dy * ball.dy > 0 &&
      this.prediction.since < this.level.aiReaction
    ) {
      this.prediction.since += dt;
    } else {
      const pt = ball.intercept(
        {
          left: this._left,
          right: this._right,
          top: -10000,
          bottom: 10000,
        },
        ball.dx * 10,
        ball.dy * 10
      );

      if (pt) {
        const t = this.minY + ball.radius;
        const b = this.maxY + this.height - ball.radius;

        while (pt.y < t || pt.y > b) {
          if (pt.y < t) {
            pt.y = t + (t - pt.y);
          } else if (pt.y > b) {
            pt.y = t + (b - t) - (pt.y - b);
          }
        }
        this.prediction = {
          since: 0,
          dx: ball.dx,
          dy: ball.dy,
          x: pt.x,
          y: pt.y,
          radius: ball.radius,
          exactX: pt.x,
          exactY: pt.y,
          d: pt.d,
        };
        const closeness =
          (ball.dx < 0 ? ball.x - this._right : this._left - ball.x) /
          this.width;
        const error = this.level.aiError * closeness;
        this.prediction.y = this.prediction.y + Engine.random(-error, error);
      } else {
        this.prediction = null;
      }
    }
  }

  moveUp() {
    this._up = 1;
  }

  moveDown() {
    this._down = 1;
  }

  stopMovingUp() {
    this._up = 0;
  }

  stopMovingDown() {
    this._down = 0;
  }

  draw(ctx: CanvasRenderingContext2D) {
    ctx.fillStyle = Colors.walls;
    ctx.fillRect(this.x, this.y, this.paddleWidth, this.paddleHeight);
    if (this.prediction && this.showPredications) {
      ctx.strokeStyle = Colors.predictionExact;
      ctx.strokeRect(
        this.prediction.x - this.prediction.radius,
        this.prediction.exactY - this.prediction.radius,
        this.prediction.radius * 2,
        this.prediction.radius * 2
      );
      ctx.strokeStyle = Colors.predictionGuess;
      ctx.strokeRect(
        this.prediction.x - this.prediction.radius,
        this.prediction.y - this.prediction.radius,
        this.prediction.radius * 2,
        this.prediction.radius * 2
      );
    }
  }
}
