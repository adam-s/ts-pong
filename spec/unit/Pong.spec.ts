import 'jest-canvas-mock';
import { Pong, Ball, Court, Menu, Paddle } from '../../src/pong';

jest.mock('../../src/pong/Ball');
jest.mock('../../src/pong/Court');
jest.mock('../../src/pong/Menu');
jest.mock('../../src/pong/Paddle');

describe('Pong', function () {
  it('should exist', function () {
    expect(Pong).toBeDefined;
  });

  let implementation: Pong;
  beforeEach(() => {
    jest.mock('../../src/pong/resources.ts', () => {
      return {
        loader: jest.fn(),
        images: {
          press1: { image: new Image() },
          press2: { image: new Image() },
          winner: { image: new Image() },
        },
      };
    });
    implementation = new Pong();
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('creates an instance', () => {
    expect(implementation).toBeDefined();
    expect(Ball).toHaveBeenCalled();
    expect(Court).toHaveBeenCalled();
    expect(Menu).toHaveBeenCalled();
    expect(Paddle).toHaveBeenCalledTimes(2);
  });

  it('starts with 0', () => {
    implementation.start(0);
    expect((Paddle as jest.Mock).mock.instances[0].setAuto).toBeCalledWith(
      true,
      8
    );
    expect((Paddle as jest.Mock).mock.instances[1].setAuto).toBeCalledWith(
      true,
      8
    );
  });

  it('starts with 1', () => {
    implementation.start(1);
    expect((Paddle as jest.Mock).mock.instances[0].setAuto).toBeCalledWith(
      false,
      8
    );
    expect((Paddle as jest.Mock).mock.instances[1].setAuto).toBeCalledWith(
      true,
      8
    );
  });

  it('starts with 2', () => {
    implementation.start(2);
    expect((Paddle as jest.Mock).mock.instances[0].setAuto).toBeCalledWith(
      false,
      8
    );
    expect((Paddle as jest.Mock).mock.instances[1].setAuto).toBeCalledWith(
      false,
      8
    );
  });

  it('stopping does nothing if not playing', () => {
    implementation['_playing'] = false;
    implementation.stop();
    expect((Paddle as jest.Mock).mock.instances[0].setAuto).not.toBeCalled();
    expect((Paddle as jest.Mock).mock.instances[1].setAuto).not.toBeCalled();
  });

  it('stops game', () => {
    implementation['_playing'] = true;
    implementation.stop();
    expect((Paddle as jest.Mock).mock.instances[0].setAuto).toBeCalledWith(
      false
    );
    expect((Paddle as jest.Mock).mock.instances[1].setAuto).toBeCalledWith(
      false
    );
  });

  it('sets the level correctly', () => {
    implementation['_scores'] = [0, 0];
    let level = implementation.level(0);
    expect(level).toBe(8);
    implementation['_scores'] = [7, 2];
    level = implementation.level(1);
    expect(level).toBe(3);
  });

  it('goaaaaaaals', () => {
    const stopSpy = jest.spyOn(implementation, 'stop');
    implementation.goal(0);
    expect((Paddle as jest.Mock).mock.instances[0].setLevel).toBeCalledWith(9);
    expect((Paddle as jest.Mock).mock.instances[1].setLevel).toBeCalledWith(7);
    expect((Ball as jest.Mock).mock.instances[0].reset).toBeCalledWith(0);
    expect(
      (Menu as jest.Mock).mock.instances[0].declareWinner
    ).not.toBeCalled();
    expect(stopSpy).not.toBeCalled();
  });

  it('goals and declares a winner', () => {
    const stopSpy = jest.spyOn(implementation, 'stop');
    implementation['_scores'] = [8, 8];
    implementation.goal(0);
    expect((Paddle as jest.Mock).mock.instances[0].setLevel).not.toBeCalled();
    expect((Paddle as jest.Mock).mock.instances[1].setLevel).not.toBeCalled();
    expect((Ball as jest.Mock).mock.instances[0].reset).not.toBeCalled();
    expect((Menu as jest.Mock).mock.instances[0].declareWinner).toBeCalledWith(
      0
    );
    expect(stopSpy).toBeCalled();
  });

  it('updates paddle when not playing', () => {
    implementation['_playing'] = false;
    implementation.update(0);
    expect((Paddle as jest.Mock).mock.instances[0].update).toBeCalledWith(
      0,
      implementation['_ball']
    );
    expect((Paddle as jest.Mock).mock.instances[1].update).toBeCalledWith(
      0,
      implementation['_ball']
    );
  });

  it('updates ball when playing', () => {
    implementation['_playing'] = true;
    implementation.update(0);
    expect((Ball as jest.Mock).mock.instances[0].update).toBeCalledWith(
      0,
      (Paddle as jest.Mock).mock.instances[0],
      (Paddle as jest.Mock).mock.instances[1]
    );
  });

  it('updates player 0 goal if ball left side passes court width', () => {
    const goalSpy = jest.spyOn(implementation, 'goal');
    implementation['_playing'] = true;
    // https://stackoverflow.com/a/41803821/494664
    // Use Object.defineProperty to change instances created by the consumer class
    // The values are destroyed if the consumer is recreated before each test case
    Object.defineProperty(implementation['_ball'], 'left', {
      get: () => 641,
    });
    implementation.update(0);
    expect(goalSpy).toBeCalledWith(0);
  });

  it('updates player 1 goal if ball right side is less than court start', () => {
    const goalSpy = jest.spyOn(implementation, 'goal');
    implementation['_playing'] = true;
    Object.defineProperty(implementation['_ball'], 'right', {
      get: () => -1,
    });
    implementation.update(0);
    expect(goalSpy).toBeCalledWith(1);
  });

  it('renders to canvas not playing', () => {
    implementation['_playing'] = false;
    const canvas = document.createElement('canvas');
    const ctx = canvas.getContext('2d');
    implementation.draw(ctx);
    const events = ctx.__getEvents();
    expect(events).toMatchSnapshot();
    ctx.__clearEvents();
  });

  it('renders to canvas when playing', () => {
    implementation['_playing'] = true;
    const canvas = document.createElement('canvas');
    const ctx = canvas.getContext('2d');
    implementation.draw(ctx);
    const events = ctx.__getEvents();
    expect(events).toMatchSnapshot();
    ctx.__clearEvents();
  });

  describe('keydown events', () => {
    it('Digit0 starts auto play', () => {
      const spy = jest
        .spyOn(implementation, 'startDemo')
        .mockImplementation(() => undefined);
      const event = new KeyboardEvent('keydown', { key: 'Digit0' });
      implementation.onkeydown(event);
      expect(spy).toBeCalled();
    });

    it('Digit1 starts single player game', () => {
      const spy = jest
        .spyOn(implementation, 'startSinglePlayer')
        .mockImplementation(() => undefined);
      const event = new KeyboardEvent('keydown', { key: 'Digit1' });
      implementation.onkeydown(event);
      expect(spy).toBeCalled();
    });

    it('Digit2 starts double player game', () => {
      const spy = jest
        .spyOn(implementation, 'startDoublePlayer')
        .mockImplementation(() => undefined);
      const event = new KeyboardEvent('keydown', { key: 'Digit2' });
      implementation.onkeydown(event);
      expect(spy).toBeCalled();
    });

    it('Escape stops game', () => {
      const spy = jest
        .spyOn(implementation, 'stop')
        .mockImplementation(() => undefined);
      const event = new KeyboardEvent('keydown', { key: 'Escape' });
      implementation.onkeydown(event);
      expect(spy).toBeCalledWith(true);
    });

    it('KeyQ moves left paddle up', () => {
      Object.defineProperty(implementation['_leftPaddle'], 'auto', {
        get: () => false,
      });
      const event = new KeyboardEvent('keydown', { key: 'KeyQ' });
      implementation.onkeydown(event);
      expect((Paddle as jest.Mock).mock.instances[0].moveUp).toBeCalled();
      expect((Paddle as jest.Mock).mock.instances[0].moveDown).not.toBeCalled();
    });

    it('KeyA moves left paddle down', () => {
      Object.defineProperty(implementation['_leftPaddle'], 'auto', {
        get: () => false,
      });
      const event = new KeyboardEvent('keydown', { key: 'KeyA' });
      implementation.onkeydown(event);
      expect((Paddle as jest.Mock).mock.instances[0].moveUp).not.toBeCalled();
      expect((Paddle as jest.Mock).mock.instances[0].moveDown).toBeCalled();
    });

    it('KeyP moves right paddle up', () => {
      Object.defineProperty(implementation['_rightPaddle'], 'auto', {
        get: () => false,
      });
      const event = new KeyboardEvent('keydown', { key: 'KeyP' });
      implementation.onkeydown(event);
      expect((Paddle as jest.Mock).mock.instances[1].moveUp).toBeCalled();
      expect((Paddle as jest.Mock).mock.instances[1].moveDown).not.toBeCalled();
    });

    it('KeyL moves right paddle down', () => {
      Object.defineProperty(implementation['_rightPaddle'], 'auto', {
        get: () => false,
      });
      const event = new KeyboardEvent('keydown', { key: 'KeyL' });
      implementation.onkeydown(event);
      expect((Paddle as jest.Mock).mock.instances[1].moveUp).not.toBeCalled();
      expect((Paddle as jest.Mock).mock.instances[1].moveDown).toBeCalled();
    });
  });

  describe('keyup events', () => {
    it('KeyQ stops left paddle moving up', () => {
      Object.defineProperty(implementation['_leftPaddle'], 'auto', {
        get: () => false,
      });
      const event = new KeyboardEvent('keyup', { key: 'KeyQ' });
      implementation.onkeyup(event);
      expect((Paddle as jest.Mock).mock.instances[0].stopMovingUp).toBeCalled();
      expect(
        (Paddle as jest.Mock).mock.instances[0].stopMovingDown
      ).not.toBeCalled();
    });

    it('KeyA stops left paddle moving down', () => {
      Object.defineProperty(implementation['_leftPaddle'], 'auto', {
        get: () => false,
      });
      const event = new KeyboardEvent('keyup', { key: 'KeyA' });
      implementation.onkeyup(event);
      expect(
        (Paddle as jest.Mock).mock.instances[0].stopMovingUp
      ).not.toBeCalled();
      expect(
        (Paddle as jest.Mock).mock.instances[0].stopMovingDown
      ).toBeCalled();
    });

    it('KeyP stops right paddle moving up', () => {
      Object.defineProperty(implementation['_rightPaddle'], 'auto', {
        get: () => false,
      });
      const event = new KeyboardEvent('keyup', { key: 'KeyP' });
      implementation.onkeyup(event);
      expect((Paddle as jest.Mock).mock.instances[1].stopMovingUp).toBeCalled();
      expect(
        (Paddle as jest.Mock).mock.instances[1].stopMovingDown
      ).not.toBeCalled();
    });

    it('KeyL stops right paddle moving down', () => {
      Object.defineProperty(implementation['_rightPaddle'], 'auto', {
        get: () => false,
      });
      const event = new KeyboardEvent('keyup', { key: 'KeyL' });
      implementation.onkeyup(event);
      expect(
        (Paddle as jest.Mock).mock.instances[1].stopMovingUp
      ).not.toBeCalled();
      expect(
        (Paddle as jest.Mock).mock.instances[1].stopMovingDown
      ).toBeCalled();
    });
  });

  it('shows footprints', () => {
    implementation.showFootprints(true);
    expect((Ball as jest.Mock).mock.instances[0].showFootprints).toBe(true);
  });

  it('shows predications', () => {
    implementation.showPredictions(true);
    expect((Paddle as jest.Mock).mock.instances[1].showPredications).toBe(true);
    expect((Paddle as jest.Mock).mock.instances[1].showPredications).toBe(true);
  });

  it('plays sounds', () => {
    implementation.playSounds(true);
    expect(implementation['_playSounds']).toBe(true);
  });
});
