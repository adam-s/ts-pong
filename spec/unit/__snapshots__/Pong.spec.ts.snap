// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pong renders to canvas not playing 1`] = `Array []`;

exports[`Pong renders to canvas when playing 1`] = `Array []`;
